### PR TITLE
Refactored WP_Debug_Data::debug_data() Part 1

### PR DESCRIPTION
**Pull request by apermo**
Changed: Moving $info['wp-filesystem'] to own function

---

Follow up to https://github.com/WordPress/wordpress-develop/pull/7027 as discussed with @dmsnell and @costdev 

To keep the structure intact over all PRs, we will start moving content from the `$info` array into own function from the bottom up, this way the result will always be identical and combarable.

Summary of the coming steps:

1. **wp-filesystem** (we are here)
2. wp-constants
3. wp-database
4. wp-server
5. wp-media
6. wp-plugins-active & wp-plugins-inactive
7. wp-mu-plugins
8. wp-themes-inactive
9. wp-parent-theme
10. wp-active-theme
11. wp-dropins
12. wp-paths-sizes
13. wp-core

Refactoring WP_Debug_Data::debug_data(); into smaller functions for

* better maintainability
* reduced complexity
* preparation for future unit tests
* improved extensibility (see trac for details)
* added php7.2 style type hints
* fixed an error with `$parent_theme_auto_update_string` showing the value from `$auto_updates_string` (This will be handled only in Step 9)

Trac ticket: https://core.trac.wordpress.org/ticket/61648#ticket

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by apermo** [comment]
I decided to move the `add_filter()` calls from `WP_Debug_Data::debug_data()` to a new function `WP_Debug_Data::init_filters()` which is called on `admin_init` this way the filters be altered in a non hacky way by plugin authors easily.

**Comment by apermo** [comment]
Update: I had to move that around a bit, in order to minimize dependencies. To allow plugin authors a convenient way to rearrange debug information, I introduced a new action `pre_debug_information` that will be fired directly before the filter `debug_information`, I decided to do this, instead of relying on plugin authors to use the filter to achieve the same, since I understand that it is a kind of hacky approach to use a filter to do an actions job.

**Comment by apermo** [comment]
In branch [refactor-wp-debug-data-old](https://github.com/apermo/wordpress-develop/tree/refactor-wp-debug-data-old) is the original PR, with a minor update, to integrate changed approaches and more phpDoc from the new refactoring.

In [refactor-wp-debug-data/all-parts](https://github.com/apermo/wordpress-develop/tree/refactor-wp-debug-data/all-parts) there is already one commit for each refactored function prepared as stated above.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (costdev) [inline src/wp-admin/includes/class-wp-debug-data.php]: I don't think using filters is the right way to go here. This, for example, would mean data that isn't currently removable will now be removable via `remove_filter()`.

What about introducing a `private` `$debug_data` property, which can be populated in `::debug_data()` by calling the new methods? The first PR could introduce the property, along with a new method to populate the filesystem data. The later PRs can then abstract the remaining logic into new methods and add `$this->data['somekey'] = self::some_method();`, for example.

Pinging @dmsnell for thoughts on this too.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (costdev) [inline src/wp-admin/includes/class-wp-debug-data.php]:
```suggestion
	public static function debug_data() {
```
Let's keep the changes strictly scoped to the purpose of the ticket, which is to reduce the complexity of `::debug_data()` and abstract its internal logic into new methods.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (costdev) [inline src/wp-admin/includes/class-wp-debug-data.php]: I don't think this action should be introduced as part of this ticket. IMO, we should focus the ticket on splitting the method up for easier maintenance, and discuss extensibility in another ticket.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (costdev) [inline src/wp-admin/includes/class-wp-debug-data.php]:
```suggestion
	 * Gets the filesystem section of the debug data.
	 *
```
This could just be a getter method, called in `::debug_data()` something like I mentioned above.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (costdev) [inline src/wp-admin/includes/class-wp-debug-data.php]:
```suggestion
	private static function wp_filesystem( array $info ): array {
```
At this stage, the new methods should begin life as `private`, given that once they're `protected` or `public`, they must remain as such for backward compatibility purposes.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: I tend to disagree with your point that it can be removed via `remove_filter()`.

Firstly you could still remove the information added by using 

```
add_filter( ...
function( $debug_data ) {
   unset( $debug_data['wp-filesystem'];
   return $debug_data;
} );
```

And I stumpled upon this trying to inject our custom debug information in a prominent position. That's why I proposed using add_filter() here, and also adding the new action. That would allow easy injection before, in the middle and after into `$debug_data`, that being said, I tend to follow your argument, that functions that are visible need to remain for backwards compatibility, this argument convinces me to follow your overall suggestion. 

Unless of course my thoughts changed your mind.

One final thought where I would love to hear your thoughts. Regarding the possibility of adding unit tests in the future. What are your thoughts regarding this? That was one of the two main reasons, why I opted to use the `add_filter()` approach. I considered having the function writing a huge array like now, or updating a class variable as high complexity for possible unit testing.

That and the possibility for more flexibility was one reason for the `add_filter()` approach.

I'd like to hear your thoughts on that before I implement your suggestions.

Comment (costdev) [reply src/wp-admin/includes/class-wp-debug-data.php]: My mistake, you're right that the data can be removed currently.

I don't have strong feelings about the use of `add_filter()`. It does however mean that we'll need to maintain each of the new `public` methods under the BC promise.

Can you share your thoughts on having `::init_filters()` now that it's not being hooked to `admin_init`?

Regarding the introduction of a new action, I still think this should be considered as part of its own ticket. Trying to cover too many problems into one "everything" ticket makes for a busy and confusing history to track. As this isn't a project, an "umbrella" ticket doesn't seem necessary, so I think we could just have one ticket per issue (abstraction, bugfix, extensibility). Abstracting a large method's logic into separate methods for ease of maintenance is a nicely scoped problem for this first ticket. We can then follow up in later tickets with bugfix, additional changes, now that the class will be more maintainable. This also means the bugfix can be reported and reproduced, and a solution tested, in a single, focused ticket.

For unit testing, if the new methods are `public`, these methods will need to be tested in isolation, in addition to `::debug_data()`. If they are `private` and called in `::debug_data()`, each section (filesystem, etc.) can have its own test method(s), which call `::debug_data()` since this would remain the entrypoint of the public API.

For example:

```php
public function test_debug_data_should_include_filesystem() {
    $debug_data = WP_Debug_Data::debug_data();
    $this->assertIsArray( $debug_data, 'Debug data should be an array.' );
    $this->assertArrayHasKey( 'wp-filesystem', $debug_data, 'The filesystem section should be included with the key "wp-filesystem".' );
    $this->assertIsArray( $debug_data['wp-filesystem'], 'The filesystem section should be an array.' );

    // Fields.
    $this->assertArrayHasKey( 'wordpress', $debug_data['wp-filesystem'], 'The WordPress section should be included.' );
    // Value testing...

    // Other fields...
}
```

The above is just an example of one approach that can work regardless of the visibility of the new methods.

While not ideal, there is also the Reflection API to make `private` methods accessible for the purpose of testing them directly. I'm just including it as a technical possibility though, rather than recommending it.

Comment (dmsnell) [reply src/wp-admin/includes/class-wp-debug-data.php]: thanks @costdev. wow. I love these kinds of refactors. Initially I think what you said is great, it's a big enough help simply to start splitting out these sections and modularizing by components that we don't have to accomplish everything in this patch or series of patched.

As I think more about it I feel less strongly about whether the filter is appropriate. The filter is _an existing filter_ so this would push Core to use its own abstractions. On the other hand, it seems like we can accomplish the goal of this refactor simply be replacing the big chunk of filesystem code with this:

```php
$info['wp-filesystem'] = static::wp_filesystem_debug_info();
```

One impact of filtering here is figuring out the primacy of this data and the expectation of whether it's there or filterable. Every filter will be provided the list of already-collected debug information and be free to modify it. Perhaps they will make some logic check based on whether it's already there.

My instinct suggests that there's a set of Core debug information people expect to be there, and it would be good to collect all of that up-front and rely on the existing `apply_filters( 'debug_information', $info )` call to pass this on to plugins.

@apermo was there something about adding your own data not already adequately served by `debug_information`?

Also, gotta love that the method name is `debug_data()`, the variable name is `$info`, and the filter's name is `debug_information` 🤦‍♂️

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: Agreed, will remove this.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: > Also, gotta love that the method name is debug_data(), the variable name is $info, and the filter's name is debug_information 🤦‍♂️

Starting with the last, since thats an internal we can rename `$info` to `$debug_information` unless anyone of you has a reason for not doing so?

Thinking through what you guys said makes me come to this scenario.

Developer A wrote code to inject his own debug information(let's say IO speed) to `wp-filesystem`, and used for whatever reason priority 0. 
So with us using the filter as well but adding our information at prio 9, we would overwrite his custom information. Merging would be possibly but not reasonable at this point.

Since we have to assume that some website maintainer will have done so in some private repo, we should go with @dmsnell 's suggestion.

That would still leave the door open to change this later to a filter approach, we can start by making the functions private. The more I think about it, the more I'm convinced that this is the right way to start. We will break down the function, prevent additional public functions from being added, and still leave the door open for easy change. 

> Can you share your thoughts on having `::init_filters()` now that it's not being hooked to `admin_init`?

Sure, I originally had it hooked to `admin_init` in one of the parent files for site health. Pretty much using:

`add_action( 'admin_init', array( WP_Debug_Data::class, 'init_filters' )` 

This crashed since `WP_Debug_Data` was not yet available. And it felt totally wrong to load the file early. And I did not want to hard code `WP_Debug_Data` to keep the code clean.

That's why I dropped the idea to hook into `admin_init` but kept `::init_filters()`. Unless anyone of you has a better idea, I would drop `::init_filters()`.

--------

One final thought.

Refactored function sets everything?
```
$debug_information['wp-filesystem'] = self::wp_filesystem();
```

OR

Refactored function only sets the fields, label description etc remains inside `::debug_data()`
```
$debug_information['wp-filesystem'] = array(
	'label'       => __( 'Filesystem Permissions' ),
	'description' => __( 'Shows whether WordPress is able to write to the directories it needs access to.' ),
	'fields'      => self::wp_filesystem(),
);
```

OR

Prefill all labels first. And fill information later.
```
$debug_information['wp-filesystem'] = array(
	'label'       => __( 'Filesystem Permissions' ),
	'description' => __( 'Shows whether WordPress is able to write to the directories it needs access to.' ),
	'fields'      => array(),
);

// Other code.

$debug_information['wp-filesystem']['fields'] = self::wp_filesystem();

// The original Filter comes below.
```

The more I look at it, the more I like the 3rd one.
* The prefill of the labels could be done by a new function at a later point, which could be super easily tested and could provide further extensibility.
* It would be pretty clean and readable to have all labels and all add fields function at one place.
* All the add fields functions could possibly be replace by filters of their own at some place, providing an option to further break down the upcoming functions.

And now that I've written all this down, I remember why I initially went for the filter approach. `wp-plugins-active` and `wp-plugins-inactive` are massively tied together, that's why I left the two of them inside a single function, and have `add_filter()` do all the work.

But I like the last idea the most, and I'm sure we will come up with a good solution for the plugins in Step 6.

Your thoughts? If you guys agree I would take care of that later today.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: Will change it but rephrase your suggestion to be conform with WPCS.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (costdev) [reply src/wp-admin/includes/class-wp-debug-data.php]: > Starting with the last, since thats an internal we can rename `$info` to `$debug_information` unless anyone of you has a reason for not doing so?

Yeah this makes sense. Technically, renaming internal variables isn't within the scope of this ticket, more something that can be done as a commit on the umbrella Coding Standards ticket for a given release (similar to [this commit](https://core.trac.wordpress.org/changeset/56411)).

> That's why I dropped the idea to hook into admin_init but kept ::init_filters(). Unless anyone of you has a better idea, I would drop ::init_filters().

Agreed - The original purpose for abstracting this no longer exists.

> One final thought.

Excellent questions. I'm still in the "thinking" phase at the moment, but for ease of maintenance and extensibility, at the moment I'm thinking of:

```php
public static function debug_data() {
	$info = array(
		'wp-core'             => self::get_wp_core(),
		'wp-paths-sizes'      => self::get_wp_paths_sizes(),
		'wp-dropins'          => self::get_wp_dropins(),
		'wp-active-theme'     => self::get_wp_active_theme(),
		'wp-parent-theme'     => self::get_wp_parent_theme(),
		'wp-themes-inactive'  => self::get_wp_themes_inactive(),
		'wp-mu-plugins'       => self::get_wp_mu_plugins(),
		'wp-media'            => self::get_wp_media(),
		'wp-server'           => self::get_wp_server(),
		'wp-database'         => self::get_wp_database(),
		'wp-constants'        => self::get_wp_constants(),
		'wp-filesystem'       => self::get_wp_filesystem(),
	);

	return apply_filters( 'debug_information', $info );
}
```
Where one of the new methods would look something like this:
```php
private static function get_wp_filesystem() {
	$upload_dir                     = wp_upload_dir();
	$is_writable_abspath            = wp_is_writable( ABSPATH );
	$is_writable_wp_content_dir     = wp_is_writable( WP_CONTENT_DIR );
	$is_writable_upload_dir         = wp_is_writable( $upload_dir['basedir'] );
	$is_writable_wp_plugin_dir      = wp_is_writable( WP_PLUGIN_DIR );
	$is_writable_template_directory = wp_is_writable( get_theme_root( get_template() ) );
	$is_writable_fonts_dir          = wp_is_writable( wp_get_font_dir()['basedir'] );

	$wp_filesystem = array(
		'label'       => __( 'Filesystem Permissions' ),
		'description' => __( 'Shows whether WordPress is able to write to the directories it needs access to.' ),
		'fields'      => array(
			'wordpress'  => array(
				'label' => __( 'The main WordPress directory' ),
				'value' => ( $is_writable_abspath ? __( 'Writable' ) : __( 'Not writable' ) ),
				'debug' => ( $is_writable_abspath ? 'writable' : 'not writable' ),
			),
			'wp-content' => array(
				'label' => __( 'The wp-content directory' ),
				'value' => ( $is_writable_wp_content_dir ? __( 'Writable' ) : __( 'Not writable' ) ),
				'debug' => ( $is_writable_wp_content_dir ? 'writable' : 'not writable' ),
			),
			'uploads'    => array(
				'label' => __( 'The uploads directory' ),
				'value' => ( $is_writable_upload_dir ? __( 'Writable' ) : __( 'Not writable' ) ),
				'debug' => ( $is_writable_upload_dir ? 'writable' : 'not writable' ),
			),
			'plugins'    => array(
				'label' => __( 'The plugins directory' ),
				'value' => ( $is_writable_wp_plugin_dir ? __( 'Writable' ) : __( 'Not writable' ) ),
				'debug' => ( $is_writable_wp_plugin_dir ? 'writable' : 'not writable' ),
			),
			'themes'     => array(
				'label' => __( 'The themes directory' ),
				'value' => ( $is_writable_template_directory ? __( 'Writable' ) : __( 'Not writable' ) ),
				'debug' => ( $is_writable_template_directory ? 'writable' : 'not writable' ),
			),
			'fonts'      => array(
				'label' => __( 'The fonts directory' ),
				'value' => ( $is_writable_fonts_dir ? __( 'Writable' ) : __( 'Not writable' ) ),
				'debug' => ( $is_writable_fonts_dir ? 'writable' : 'not writable' ),
			),
		),
	);

	return $wp_filesystem; // Super easy to add a filter for this in future if we want.
}
```
1. Keeps section label, field label, and field contents together.
2. For Core, adding a new section means adding a new array entry in `::debug_data()`, and a new method to handle all the rest.
3. For extenders, adding a new section means filtering `debug_information` to add a new array entry with that section's data.
4. PHPUnit Testing is still as straightforward.
5. In future, we can look at the possibility of introducing section-specific filters, such as `'debug_information_wp_filesystem'`, with a really easy way to introduce the filters on the `return` line.

Note:
I've intentionally left active and inactive plugins out of the example `::debug_data()` above.

I think we should carefully consider whether the logic for active and inactive plugins should impact our design decision here. Certainly there's a performance impact, as separating them will lead to additional cache hits for `get_plugins()` and such, but there's also transients like `'update_plugins'` that have no timeout, so separation would lead to additional API calls (and in super extremely rare cases, possibly even different data).

We could do something like:

```php
private static function get_wp_plugins() {
	$wp_plugins = array(
		'wp-plugins-active' => array(
			'label'      => __( 'Active Plugins' ),
			'show_count' => true,
			'fields'     => array(),
		),
		'wp-plugins-inactive' => array(
			'label'      => __( 'Inactive Plugins' ),
			'show_count' => true,
			'fields'     => array(),
		),
	);

	// Logic to assign the 'fields' keys.

	return $wp_plugins;
}

public static function debug_data() {
	list(
		// Supported in PHP 7.1+
		'wp-plugins-active'   => $wp_active_plugins,
		'wp-plugins-inactive' => $wp_inactive_plugins
	) = self::get_wp_plugins();

	// In the array:
	'wp-plugins-active'   => $wp_active_plugins,
	'wp-plugins-inactive' => $wp_inactive_plugins,
}
```
1. Retains the existing logic, but in an abstracted method.
2. Handles an outlier as an outlier, rather than having it impact the overall design.
3. Uses the same '`wp-plugins-active'` and `'wp-plugins-inactive'` keys in `::get_wp_plugins()` in case section-specific filtering is introduced later. Having `$wp_plugins` as an associative array is more readable than as an indexed array.
4. We can adjust this approach relatively freely in future due to the `private` visibility of the `get_wp_plugins()` method and lack of section-specific filtering at this time.

What do others think?

Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: Looks good to me, unless Dennis has any other ideas I'll start implementing things later, expect the update by tomorrow morning UTC.

Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: @costdev Revised version is here.

I've decided to keep one difference to your suggestion and have 
```
return array(
	'label'       => __( 'Filesystem Permissions' ),
	'description' => __( 'Shows whether WordPress is able to write to the directories it needs access to.' ),
	'fields'      => $fields,
);
```

While your suggestion would perfectly work here, the other upcoming functions will be way move complex, where that approach shortens array identifiers, and thus improves readability and I prefer the consistency over all functions here which could ultimately allow us an easy refactor to classes with an interface.

But even if we won't use that, imho that is better readable if we keep a single schema.

Hope I did not forget anything from the review.

Comment (dmsnell) [reply src/wp-admin/includes/class-wp-debug-data.php]: I like the idea of keeping the labels and the fields together, but I don't feel strongly at this point about it.

**Review by dmsnell — APPROVED** [review]
I'll let some other folks review this if they want, but the code as proposed seems clear enough to me.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: @dmsnell what do you mean? Having both in a single function (what we have now), or have both in a single array definition?

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (costdev) [inline src/wp-admin/includes/class-wp-debug-data.php]:
```suggestion
	 * Gets the file system section of the debug data.
```
Minor change: Core uses third-person singular verbs for docblock summaries("Gets" not "Get", "Sets" not "Set", etc.)

[Ref](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#documenting-tips)

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the updates @apermo!

I've left one small change to meet Docblock language standards.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: My bad. I was so sure that it was the other way round. I frequently mix that up.

**Comment by costdev** [comment]
Thanks @apermo! I think this looks like a good approach for making things more maintainable.

Hopefully @afragen and @clorith have some time to take a look at this and review as Site Health component maintainers.

**Review by costdev — APPROVED** [review]
Approving from my perspective, pending reviews from one or more component maintainers.

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (dmsnell) [reply src/wp-admin/includes/class-wp-debug-data.php]: @apermo I think what's in the code now is good. where the label and information all come from the `get_wp_filesystem_info()` method

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (dmsnell) [inline src/wp-admin/includes/class-wp-debug-data.php]:
```
 * @ since 6.7.0
```

Comment (dmsnell) [reply src/wp-admin/includes/class-wp-debug-data.php]: noting for the merge

Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: shall i add this to the PR?

Comment (dmsnell) [reply src/wp-admin/includes/class-wp-debug-data.php]: if you'd like!

Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: Done

File: src/wp-admin/includes/class-wp-debug-data.php
Comment (apermo) [reply src/wp-admin/includes/class-wp-debug-data.php]: I guess we resolved this? I'll close it, just open it back up if needed.

**Comment by afragen** [comment]
LGTM 👍

**Comment by dmsnell** [comment]
Prepping to merge.

**Comment by dmsnell** [comment]
Merged in [[58830]](https://core.trac.wordpress.org/changeset/58830)
https://github.com/wordpress/wordpress-develop/commit/345699c0711d00dc15a1f3090694c2c8a114a29b

**Pull request closed without merging**
